File: app/src/main/java/io/github/sithengineer/motoqueiro/ui/home/HomeFragment.java
package io.github.sithengineer.motoqueiro.ui.home;

import android.content.Intent;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.design.widget.TextInputLayout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import butterknife.BindView;
import com.jakewharton.rxbinding.view.RxView;
import com.jakewharton.rxbinding.widget.RxTextView;
import io.github.sithengineer.motoqueiro.BaseFragment;
import io.github.sithengineer.motoqueiro.MotoqueiroApp;
import io.github.sithengineer.motoqueiro.PermissionAuthority;
import io.github.sithengineer.motoqueiro.R;
import javax.inject.Inject;
import rx.Observable;

public class HomeFragment extends BaseFragment<HomeContract.Presenter>
    implements HomeContract.View {

  @BindView(R.id.miband_address_input_wrapper) TextInputLayout miBandAddressLayout;
  @BindView(R.id.miband_address_input) EditText miBandAddress;
  @BindView(R.id.ride_name_input_wrapper) TextInputLayout rideNameLayout;
  @BindView(R.id.ride_name_input) EditText rideName;
  @BindView(R.id.start_ride) Button startRideButton;
  @BindView(R.id.phone_position_spinner) Spinner phonePosition;
  @Inject HomeContract.Presenter presenter;

  public static HomeFragment newInstance() {
    return new HomeFragment();
  }

  @Override public int getViewId() {
    return R.layout.fragment_home;
  }

  @Override public void inject() {
    MotoqueiroApp.get(getContext())
        .createUiComponent()
        .with(new HomeModule(this, new HomeNavigator(getActivity()),
            (PermissionAuthority) getActivity()))
        .inject(this);
  }

  @Nullable @Override public HomeContract.Presenter getPresenter() {
    return presenter;
  }

  @Nullable @Override
  public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
      @Nullable Bundle savedInstanceState) {
    final View view = super.onCreateView(inflater, container, savedInstanceState);
    ArrayAdapter<CharSequence> adapter =
        ArrayAdapter.createFromResource(getActivity(), R.array.home_wheres_phone_array,
            android.R.layout.simple_spinner_item);
    adapter.setDropDownViewResource(R.layout.support_simple_spinner_dropdown_item);
    phonePosition.setAdapter(adapter);
    return view;
  }

  @Override public Observable<Void> handleStartRideClick() {
    return RxView.clicks(startRideButton);
  }

  @Override public void sendToGpsSettings() {
    startActivity(new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS));
  }

  @Override public void showActivateGpsViewMessage() {
    Snackbar.make(startRideButton, R.string.home_activate_gps, Snackbar.LENGTH_LONG).show();
  }

  @Override public void showGenericError() {
    Snackbar.make(startRideButton, R.string.home_generic_error, Snackbar.LENGTH_LONG).show();
  }

  @Override public void showMiBandAddress(String miBandAddress) {
    this.miBandAddress.setText(miBandAddress);
  }

  @Override public Observable<String> getMiBandAddressChanges() {
    return RxTextView.textChanges(miBandAddress).map(text -> text.toString());
  }

  @Override public void showMiBandAddressError() {
    miBandAddressLayout.setError(getString(R.string.home_view_heart_rate_band_address_error));
  }

  @Override public void cleanMiBandAddressError() {
    miBandAddressLayout.setError(null);
  }

  @Override public void showGivePermissionsMessage() {
    Snackbar.make(startRideButton, R.string.home_need_all_permissions, Snackbar.LENGTH_LONG).show();
  }

  @Override public String getRideName() {
    return rideName.getText().toString();
  }

  @Override public int getSelectedDevicePosition() {
    return phonePosition.getSelectedItemPosition();
  }
}

File: app/src/main/java/io/github/sithengineer/motoqueiro/ui/home/HomeFragment.kt
package io.github.sithengineer.motoqueiro.ui.home

import android.content.Intent
import android.os.Bundle
import android.provider.Settings
import android.support.design.widget.Snackbar
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import com.jakewharton.rxbinding2.view.RxView
import com.jakewharton.rxbinding2.widget.RxTextView
import io.github.sithengineer.motoqueiro.MotoqueiroApp
import io.github.sithengineer.motoqueiro.PermissionAuthority
import io.github.sithengineer.motoqueiro.R
import io.github.sithengineer.motoqueiro.ui.BaseFragment
import io.github.sithengineer.motoqueiro.ui.home.HomeContract.Presenter
import io.reactivex.Observable
import kotlinx.android.synthetic.main.fragment_home.miband_address_input
import kotlinx.android.synthetic.main.fragment_home.miband_address_input_wrapper
import kotlinx.android.synthetic.main.fragment_home.phone_position_spinner
import kotlinx.android.synthetic.main.fragment_home.ride_name_input
import kotlinx.android.synthetic.main.fragment_home.start_ride
import javax.inject.Inject

class HomeFragment : BaseFragment<Presenter>(), HomeContract.View {

  @Inject
  internal lateinit var presenter: HomeContract.Presenter

  override fun getPresenter(): Presenter? = presenter

  public override val viewId: Int
    get() = R.layout.fragment_home

  override fun miBandAddressChanges(): Observable<String> =
      RxTextView.textChanges(miband_address_input).map { text -> text.toString() }

  override fun selectedDevicePosition(): Int = phone_position_spinner.selectedItemPosition

  override fun inject() {
    MotoqueiroApp[context!!]
        .createUiComponent()
        .with(HomeModule(this, HomeNavigator(activity!!), activity as PermissionAuthority))
        .inject(this)
  }

  override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
      savedInstanceState: Bundle?): View? {
    val view = super.onCreateView(inflater, container, savedInstanceState)
    val adapter = ArrayAdapter.createFromResource(activity!!, R.array.home_wheres_phone_array,
        android.R.layout.simple_spinner_item)
    adapter.setDropDownViewResource(R.layout.support_simple_spinner_dropdown_item)
    phone_position_spinner.adapter = adapter
    return view
  }

  override fun handleStartRideClick(): Observable<Any> {
    return RxView.clicks(start_ride)
  }

  override fun sendToGpsSettings() {
    startActivity(Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS))
  }

  override fun showActivateGpsViewMessage() {
    Snackbar.make(start_ride, R.string.home_activate_gps, Snackbar.LENGTH_LONG).show()
  }

  override fun showGenericError() {
    Snackbar.make(start_ride, R.string.home_generic_error, Snackbar.LENGTH_LONG).show()
  }

  override fun showMiBandAddress(miBandAddress: String) {
    this.miband_address_input.setText(miBandAddress)
  }

  override fun showMiBandAddressError() {
    miband_address_input_wrapper.error = getString(R.string.home_view_heart_rate_band_address_error)
  }

  override fun cleanMiBandAddressError() {
    miband_address_input_wrapper.error = null
  }

  override fun showGivePermissionsMessage() {
    Snackbar.make(start_ride, R.string.home_need_all_permissions,
        Snackbar.LENGTH_LONG).show()
  }

  override fun getRideName(): String {
    return ride_name_input.text.toString()
  }

  companion object {

    fun newInstance(): HomeFragment {
      return HomeFragment()
    }
  }
}
